### PR TITLE
TAPI memory improvements for 8.2

### PR DESCRIPTION
**Pull request by 6hundreds**
Cherrypick of https://github.com/gradle/gradle/pull/25267

### CODE REVIEW

**Review by bamboo — APPROVED** [review]

**Comment by eskatos** [comment]
You mentioned missing cross version tests, I would be more confident with this change if these tests were added.

**Comment by 6hundreds** [comment]
@bot-gradle test RFR

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/CustomToolingModelCrossVersionSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            fetchCustomModelsWithRestrictedMemoryAction(connection)
+        }
+
+        then:
+        notThrown(GradleConnectionException)
+        assertHasConfigureSuccessfulLogging()
+    }
+
+    @ToolingApiVersion("<8.2")
+    def "older TAPI versions is not releasing memory during BuildAction"() {
```
Comment (donat) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/CustomToolingModelCrossVersionSpec.groovy:96]:
```suggestion
    def "older Tooling API versions cannot free memory when executing a build action"() {
```

Comment (donat) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/CustomToolingModelCrossVersionSpec.groovy:96]: Please move the test to the top of the class. That way the sequence of tests are read like:
- this is a bug in old versions
- starting from Gradel 8.2, it is fixed

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/CustomToolingModelCrossVersionSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            fetchCustomModelsWithRestrictedMemoryAction(connection)
+        }
+
+        then:
+        thrown(GradleConnectionException)
+        caughtGradleConnectionException.cause.cause instanceof OutOfMemoryError
+    }
+
+    private fetchCustomModelsWithRestrictedMemoryAction(ProjectConnection connection) {
+        connection.action()
```
Comment (donat) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/CustomToolingModelCrossVersionSpec.groovy:108]: ❓ Wouldn't this be sufficient?

```
connection.action(new FetchProjectsCustomModelsAction()).run();
```

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/FetchProjectsCustomModelsAction.groovy [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.tooling.r82
+
+import org.gradle.tooling.BuildAction
+import org.gradle.tooling.BuildController
+import org.gradle.tooling.model.gradle.GradleBuild
+
+class FetchProjectsCustomModelsAction implements BuildAction {
```
Comment (donat) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/FetchProjectsCustomModelsAction.groovy:23]:
```suggestion
class FetchProjectsCustomModelsAction implements BuildAction<Void> {
```

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/FetchProjectsCustomModelsAction.groovy [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Override
+    Object execute(BuildController controller) {
+        def projects = controller.getModel(GradleBuild).projects
+        projects.forEach { project ->
+            controller.getModel(project, CustomModel)
+        }
+
+        // Returning value is doesn't matter. Most important part is fetching custom models in a BuildAction classloader.
+        return 42
```
Comment (donat) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r82/FetchProjectsCustomModelsAction.groovy:33]:
```suggestion
```

**Review by donat — CHANGES_REQUESTED** [review]
Review for the cross-version test.

**Comment by 6hundreds** [comment]
> Review for the cross-version test.

Applied your comments

**Comment by 6hundreds** [comment]
@bot-gradle test RFR

**Comment by 6hundreds** [comment]
@bot-gradle test RFR

**Comment by 6hundreds** [comment]
@bot-gradle test this

**Comment by 6hundreds** [comment]
@bot-gradle test RFR

**Comment by 6hundreds** [comment]
@bot-gradle test and merge

**Comment by 6hundreds** [comment]
@bot-gradle cancel

**Comment by 6hundreds** [comment]
@bot-gradle test and merge

**Pull request merged**